uni/lib/unittest.icn

```diff
@@ -0,0 +1,166 @@
+# TODO: assertions
+#     - asserting objects
+#     - asserting nested values (list, table, objects, etc.)
+# TODO: Test Reporter
+#     - add message to results
+#     - add backtrace if possible
+# TODO: exit 0 when success, 1 when failure.
+# TODO: there should be a way to discover and load tests dynamically
+# TODO: add timer, see tests/unicon/tester.icn
+# TODO: how to catch stdin/stdout/stderr within tests, see capture().
+# TODO: enable/disable verbosity.
+
+package unittest
+
+link ximage
+
+import lang
+
+invocable all
+
+class TestCase : Object(__status)
+   $define STATUS_SUCCESS 0
+   $define STATUS_FAIL 1
+
+   method setupClass()
+      write(classname(self), "::setupClass()")
+   end
+
+   method teardownClass()
+      write(classname(self), "::teardownClass()")
+   end
+
+   method setup()
+      write(classname(self), "::setup()")
+   end
+
+   method teardown()
+      write(classname(self), "::teardown()")
+   end
+
+   method __reset()
+      self.__status := STATUS_SUCCESS
+   end
+
+   method assert(expression)
+      if not @expression[1] then {
+         self.__status := STATUS_FAIL
+         write("Assertion failed.")
+         fail
+      }
+
+      return
+   end
+
+   method assertEqual(expected, actual)
+      if expected ~== actual then {
+         write("Assertion failed: expected ", expected, ", but got ", actual)
+         self.__status := STATUS_FAIL
+         fail
+      }
+
+      return
+   end
+
+initially
+   /__status := STATUS_SUCCESS
+end
+
+class TestSuite(__tests, __testReporter)
+   method add(testCase)
+      cname := classname(testCase)
+      if (\cname & (not (cname == "unittest__TestCase"))) then {
+         if \testCase.__m["Type"] then {
+            every ctype := testCase.Type() do {
+               if ctype == "unittest::TestCase" then {
+                  put(__tests, testCase)
+               }
+            }
+         }
+      }
+   end
+
+   method run()
+      local mname
+
+      every test := !__tests do {
+         cname := classname(test)
+
+         test.setupClass()
+         every image(!test.__m) ? {
+            mname := (="procedure ", tab(0)) | next
+            if mname ? (tab(*cname+1), ="_", match("test")) then {
+               write(ximage(mname))
+               test.__reset()
+               test.setup()
+               mname(test)
+               test.teardown()
+
+               self.__testReporter.result(cname, mname, test.__status)
+               write()
+            }
+         }
```

Comment:
cname := classname(test)

         test.setupClass()
         every image(!test.__m) ? {
            mname := (="procedure ", tab(0)) | next
            if mname ? (tab(*cname+1), ="_", match("test")) then {
               write(ximage(mname))
               test.__reset()
               test.setup()
               mname(test)
               test.teardown()

               self.__testReporter.result(cname, mname, test.__status)
               write()
            }
         }

(1) Is the intent here to only check a 'locally'-defined method?  As written using the tab(*cname+1) only a locally defined method is found.  If there isn't one this won't find an inherited method of the same name.

(2) Shouldn't you break out of the inner every if the method is found?

Reply:
1) Any given test class is a subclass of TestCase. The setupClass,(), setup(), teardown() and teardownClass() are ran exactly where they are supposed to be ran. Subclassing TestCase actually allow you to do what you want with those methods, for example, I reset values in setup() to ensure each test start afresh. setupClass could connect you to a DB or open a file whereas  teardownClass would close them.

Subclassing a subclassed class of TestCase is a not typical scenario. Using multiple inheritance hasn't been taken into account for now. According to https://github.com/uniconproject/unicon/issues/422, we would have to sit and think what the right approach it should be, or even if we need this.

This is worth mentioning that the implementation is not mutex friendly just yet, as `tests/unicon/tester.icn` is. 

For the rest, any method starting with "test" are to be run and anything else is not. It allows the possibility to define utility methods that may be useful within some tests, called at the discretion of the programmer within the tests. 

My current approach is to provide a Unit Test Framework that is fairly conventional to other TDD frameworks. 

2) No, a test could have zero, one or many test methods. It has to search the whole __m space.